providers/line/line.go

```diff
@@ -0,0 +1,162 @@
+type Provider struct {
+	ClientKey    string
+	Secret       string
+	CallbackURL  string
+	HTTPClient   *http.Client
+	config       *oauth2.Config
+	providerName string
+}
+
+// New creates a new Line provider and sets up important connection details.
+// You should always call `line.New` to get a new provider.  Never try to
+// create one manually.
+func New(clientKey, secret, callbackURL string, scopes ...string) *Provider {
+	p := &Provider{
+		ClientKey:    clientKey,
+		Secret:       secret,
+		CallbackURL:  callbackURL,
+		providerName: "line",
+	}
+	p.config = newConfig(p, scopes)
+	return p
+}
+
+// Name is the name used to retrieve this provider later.
+func (p *Provider) Name() string {
+	return p.providerName
+}
+
+// SetName is to update the name of the provider (needed in case of multiple providers of 1 type)
+func (p *Provider) SetName(name string) {
+	p.providerName = name
+}
+
+// Client returns a pointer to http.Client setting some client fallback.
+func (p *Provider) Client() *http.Client {
+	return goth.HTTPClientWithFallBack(p.HTTPClient)
+}
+
+// Debug is a no-op for the line package.
+func (p *Provider) Debug(debug bool) {}
+
+// BeginAuth asks line.me for an authentication end-point.
+func (p *Provider) BeginAuth(state string) (goth.Session, error) {
+	return &Session{
+		AuthURL: p.config.AuthCodeURL(state),
+	}, nil
+}
+
+// FetchUser will go to line.me and access basic information about the user.
+func (p *Provider) FetchUser(session goth.Session) (goth.User, error) {
+	sess := session.(*Session)
+	user := goth.User{
+		AccessToken:  sess.AccessToken,
+		Provider:     p.Name(),
+		RefreshToken: sess.RefreshToken,
+		ExpiresAt:    sess.ExpiresAt,
+	}
+
+	if user.AccessToken == "" {
+		// data is not yet retrieved since accessToken is still empty
+		return user, fmt.Errorf("%s cannot get user information without accessToken", p.providerName)
+	}
+
+	// Get the userID, line needs userID in order to get user profile info
+	c := p.Client()
+	req, err := http.NewRequest("GET", endpointUser, nil)
+	if err != nil {
+		return user, err
+	}
+
+	req.Header.Add("Authorization", "Bearer "+sess.AccessToken)
+
+	response, err := c.Do(req)
+	if err != nil {
+		if response != nil {
+			response.Body.Close()
+		}
+		return user, err
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return user, fmt.Errorf("%s responded with a %d trying to fetch user information", p.providerName, response.StatusCode)
+	}
+
+	bits, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return user, err
+	}
+
+	u := struct {
+		Name          string `json:"name"`
+		UserID        string `json:"userId"`
+		DisplayName   string `json:"displayName"`
+		PictureURL    string `json:"pictureUrl"`
+		StatusMessage string `json:"statusMessage"`
+	}{}
+
+	if err = json.NewDecoder(bytes.NewReader(bits)).Decode(&u); err != nil {
+		return user, err
+	}
+
+	name := strings.Split(u.DisplayName, " ")
```

Comment:
This line needs an improvement, what do you think @bentranter? Don't have any idea how to properly handle splitting strings though 😅. From the docs of [LINE](https://developers.line.biz/en/docs/social-api/getting-user-profiles/) the display name is a one line string, so I need to split it to get the First and Last Name.

Reply:
This seems ok to me, although I'm worried about the possible case where `u.DisplayName` is only one word long, so that `strings.Split(u.DisplayName, " ")` returns a slice containing a single string. In that case, trying to access `name[1]` would panic.

It might be simpler not to bother setting the first and last name (since LINE doesn't seem to support those fields explicitly), and just setting the `u.DisplayName` to the nickname field. I also wouldn't bother changing the display name to lowercase before setting it to nickname, in case a user of this provider wants to compare display names as-is.

Reply:
> returns a slice containing a single string. In that case, trying to access name[1] would panic.

- Yeah, it will really panic. But now, I won't bother setting it :) thanks for the feedback 👍 

> I also wouldn't bother changing the display name to lowercase before setting it to nickname, in case a user of this provider wants to compare display names as-is.

- On it. Will update my PR to apply all the reviewed changes :)

Thanks @bentranter 🍻